File: 01-beginner/02-ros2-core-concepts/demo/python/talker.py
#!/usr/bin/env python3
"""Publishes a counting string message on /chatter once per second."""

import rclpy
from rclpy.node import Node
from std_msgs.msg import String


class Talker(Node):
    def __init__(self):
        # "talker" is this node's name — it's what shows up in
        # `ros2 node list` while this is running.
        super().__init__("talker")

        # create_publisher(msg_type, topic_name, queue_size).
        # queue_size (10 here) is how many outgoing messages ROS2 will
        # buffer if a subscriber can't keep up — irrelevant at 1 Hz, but
        # it's a required argument so we pick a small, conventional value.
        self._publisher = self.create_publisher(String, "chatter", 10)

        # A count we increment each publish, just so the message content
        # visibly changes each time — makes it obvious in the listener's
        # output that new messages are actually arriving, not the same
        # one being re-printed.
        self._count = 0

        # create_timer(period_seconds, callback) — the executor calls
        # self._publish() every 1.0 seconds once we start spinning below.
        # This is "push" style: the node acts on its own schedule, it
        # doesn't wait to be asked for data.
        self._timer = self.create_timer(1.0, self._publish)

    def _publish(self):
        msg = String()
        msg.data = f'Hello, ROS2! count={self._count}'
        self._publisher.publish(msg)
        self.get_logger().info(f"Publishing: '{msg.data}'")
        self._count += 1


def main():
    # rclpy.init() must run before any Node is constructed — it sets up
    # the underlying communication layer (DDS) this process will use to
    # talk to the rest of the ROS2 graph.
    rclpy.init()
    node = Talker()
    try:
        # spin() blocks here, handing control to the executor, which
        # invokes our timer callback repeatedly until Ctrl+C.
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        # Explicit cleanup: release the node's resources and shut down
        # the communication layer cleanly rather than relying on process
        # exit to do it implicitly.
        node.destroy_node()
        rclpy.shutdown()


if __name__ == "__main__":
    main()

File: 01-beginner/02-ros2-core-concepts/demo/cpp/src/talker.cpp
// Publishes a counting string message on /chatter once per second.
//
// This is the C++ equivalent of demo/python/talker.py — same pattern
// (node -> publisher -> timer callback), different language. Compare the
// two side by side to see what's language-specific (memory management,
// syntax) vs. what's the actual ROS2 concept (publisher, timer, spin).

#include <chrono>
#include <memory>

#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/string.hpp"

using namespace std::chrono_literals;  // lets us write "1s" for 1 second

class Talker : public rclcpp::Node
{
public:
  Talker() : Node("talker"), count_(0)
  {
    // create_publisher<MsgType>(topic_name, queue_size) — same shape as
    // rclpy's create_publisher, just templated on the message type
    // instead of taking it as a positional argument.
    publisher_ = this->create_publisher<std_msgs::msg::String>("chatter", 10);

    // create_wall_timer binds this->publish_message() as the callback,
    // fired every 1 second. `this` is captured so the lambda can call
    // back into the node's members.
    timer_ = this->create_wall_timer(
      1s, std::bind(&Talker::publish_message, this));
  }

private:
  void publish_message()
  {
    auto msg = std_msgs::msg::String();
    msg.data = "Hello, ROS2! count=" + std::to_string(count_);
    RCLCPP_INFO(this->get_logger(), "Publishing: '%s'", msg.data.c_str());
    publisher_->publish(msg);
    count_++;
  }

  rclcpp::TimerBase::SharedPtr timer_;
  rclcpp::Publisher<std_msgs::msg::String>::SharedPtr publisher_;
  size_t count_;
};

int main(int argc, char * argv[])
{
  // rclcpp::init is the C++ equivalent of rclpy.init() — sets up the
  // communication layer before any node is constructed.
  rclcpp::init(argc, argv);

  // spin() blocks here and drives the timer callback until Ctrl+C
  // (rclcpp::shutdown gets called internally on SIGINT).
  rclcpp::spin(std::make_shared<Talker>());

  rclcpp::shutdown();
  return 0;
}
